### PR TITLE
Refactor tests

### PR DESCRIPTION
**Pull request by qubvel**
The idea is to:
1) make tests a bit more modular and flexible
2) speedup
3) improve coverage

### CODE REVIEW

**Comment by adamjstewart** [comment]
Can you integrate codecov so we can actually measure codecov on every PR?

**Comment by qubvel** [comment]
Never used it before but can have a look :) Did you mean `pytest-cov`? Do you have any good examples for CI jobs?

**Comment by adamjstewart** [comment]
`pytest-cov` let's you calculate and report coverage, then you can create an account and upload it to https://about.codecov.io/ in CI. https://docs.codecov.com/docs/github-tutorial gives a good explanation of the steps you need to do. I've done it for a few projects and can help you get started if you get stuck.

**Comment by qubvel** [comment]
Thanks! Looks like all set 👍

**Comment by qubvel** [comment]
Merging it for now, test-cov can be improved in follow-up PRs. The most important I think is to have full-pretrained model tests to ensure we do not break anything with updates, it's added as a separate job "logits_match". Coverage can be improved for loss and metrics in the follow-up PR or in case of new updates in these modules.

File: requirements/minimum.old [hunk 1]
```diff
@@ -8,3 +8,4 @@ timm==0.9.0
 torch==1.9.0
 torchvision==0.10.0
 tqdm==4.42.1
+Jinja2==3.0.0
```
Comment (adamjstewart) [inline requirements/minimum.old:11]: Why is Jinja2 required? I don't see any code that uses it

File: tests/encoders/base.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
+import unittest
```
Comment (adamjstewart) [inline tests/encoders/base.py:1]: Why use unittest instead of pytest? unittest is more or less obsolete

File: tests/utils.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import os
+import timm
+import torch
+import unittest
+
+from packaging.version import Version
```
Comment (adamjstewart) [inline tests/utils.py:6]: Seems like we have a new test dependency on packaging, this should be added to `pyproject.toml` and `requirements/tests.txt`

File: tests/utils.py
Comment (adamjstewart) [inline tests/utils.py:1]: I would personally use pytest fixtures for this in `conftest.py` so they are available to all test code without import

**Review by adamjstewart — COMMENTED** [review]
Still don't know what the test coverage was or is now

File: requirements/minimum.old (hunk 1 above)
Comment (qubvel) [reply requirements/minimum.old:11]: save/load pretrained raises an error on huggingface_hub 0.24 and asks to install `Jinja2`

Comment (qubvel) [reply requirements/minimum.old:11]: alternatively we can bump minimum version for `huggingface_hub`

File: tests/encoders/base.py (hunk 2 above)
Comment (qubvel) [reply tests/encoders/base.py:1]: Yeah, but I like the idea of test inheritance, don't know if we can use it with pytest. But we still use pytest to collect and run tests

File: tests/utils.py (hunk 3 above)
Comment (qubvel) [reply tests/utils.py:6]: Hm... I thought it's a default python package, but probably install with other dependency

**Comment by qubvel** [comment]
Thanks for the review!

**Comment by qubvel** [comment]
Feel free to refactor if you know a better/cleaner way to organize tests! I'm going to be mostly off for holidays up to 3rd of Jan, but this week can review/merge. Then I think it would be great to make a new release, what do you think?

File: tests/utils.py (hunk 3 above)
Comment (adamjstewart) [reply tests/utils.py:6]: It is not default, it is installed using `pip install packaging`. It may or may not be installed, we should add an explicit dependency on it.

**Comment by adamjstewart** [comment]
Yes, would be very excited for a new release so we can start using newer timm!
